### ISSUE TITLE
twisted.scripts._twistd_unix.removePID uses ancient error logging style

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2972]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2972|
|Type|defect|
|Created|2008-01-04 21:01:29Z|
|Branch|https://github.com/twisted/twisted/tree/removepid-logging-2972|

```removePID``` calls ```log.deferr``` in two places with no arguments, each place immediately after calling ```log.msg``` with an explanatory message.

These should be calls to ```log.err``` with the ```log.msg``` argument given as the a __why_ parameter.  There should be no ```log.msg``` calls.

Control flow could probably also be arranged so that there's only one call to ```log.err```.

Attachments:

* [scripts-_twistd_unix-logging.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/792/7925213d2339a1ab016c9d94edd2ff680a8f68ce/e268b1d1001677a81fa52b8dbf0c22f65aac1f58.patch) (759 bytes) - added by jesstess on 2009-08-31 22:49:40Z - no more log.deferr
* [scripts-_twistd_unix-test.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/792/7925213d2339a1ab016c9d94edd2ff680a8f68ce/8721095853af09063bec9bb21523a969128cd3c4.patch) (854 bytes) - added by jesstess on 2009-09-02 19:03:53Z - test case for removePID errors

<details><summary>Searchable metadata</summary>

```
trac-id__2972 2972
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_removepid_logging_2972 branches/removepid-logging-2972
branch_author__exarkun__jesstess exarkun, jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1199480489000000 1199480489000000
changetime__1252254887000000 1252254887000000
version__None None
owner__ 
cc__jesstess cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_1"></a>|@jesstess removed owner|
|-|-|

In the patch I followed the convention that I observed grepping through the source for other instances of log.err() that have a _why.

As an aside, I noticed that log.err() is called with just one string all over the place even though the _stuff parameter is type: NoneType, Exception, or Failure according to the docs. Say the docs, _stuff "will be wrapped in a Failure". If this example counts as "wrapping something in a failure":

from twisted.python import failure

failure.Failure("Test")

__main__:1: DeprecationWarning: Don't pass strings (like 'Test') to failure.Failure (replacing with a DefaultException).

<twisted.python.failure.Failure <class 'twisted.python.failure.DefaultException'>>



Given the DeprecationWarning, is it worth filing a ticket about getting rid of those cases?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks for the patch.  The change looks good, but it should probably be accompanied by some unit tests.  The messages themselves don't need to be tested, but there should be coverage for the exception being logged.  The way to do this is generally to provoke the error behavior and then use `TestCase.flushLoggedErrors` to inspect them.

Filing tickets for any other code that's calling `log.err` with a string would be great.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

The -test patch has a test case for the exception logged in the original patch, namely an OSError logged by UnixApplicationRunner.removePID.

[#3996](https://github.com/twisted/twisted/issues/3996) was opened to handle all instances of log.err getting just a string.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Looks good.  I'm going to address a couple style points.  Unfortunately, the Windows build slaves are mostly offline at the moment.  I want to wait until I can see this code run on Windows before declaring it okay.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [27283]) Branching to 'removepid-logging-2972'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [27284]) Apply jessjess's patches plus another test and some minor twiddling

refs [#2972](https://github.com/twisted/twisted/issues/2972)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Someone else should take a look at this now.  And hopefully the Windows slaves will come back soon.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

The copyright date(s) need to be updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27285]) Update copyright date

refs [#2972](https://github.com/twisted/twisted/issues/2972)

thanks thijs :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @exarkun|
|-|-|

Nice fix, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [27294]) Merge removepid-logging-2972

Author: jesstess, exarkun
Reviewer: exarkun, therve
Fixes: [#2972](https://github.com/twisted/twisted/issues/2972)

Switch the PID file checking code's error reporting to use the
relatively newish two-argument `log.err` form instead of the older
`log.msg`-then`log.err` style.  Also add some tests for this code.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
